#!/usr/bin/perl -w

#===============================================================================
#
#    Created: 
#    Copyright (C) 2008, University of Liverpool.
#    Author: Kevin Ashelford.
#
#    Contact details:
#    Email: k.ashelford@liv.ac.uk
#    Address:  School of Biological Sciences, University of Liverpool, 
#    Biosciences Building, Crown Street, Liverpool, UK. L69 7ZB
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#===============================================================================

# INTERFACE NAME
package Kea::IO::Newbler::Ace::IReaderHandler;
use Kea::Object;
our @ISA = qw(Kea::Object);

use strict;
use warnings;

our $_message = "Undefined method";

################################################################################

sub _nextAS 	{Kea::Object->throw($_message);} 

sub _nextCO 	{Kea::Object->throw($_message);} 

sub _nextBQ 	{Kea::Object->throw($_message);} 
	
sub _nextAF 	{Kea::Object->throw($_message);} 

sub _nextBS 	{Kea::Object->throw($_message);} 

sub _nextRD 	{Kea::Object->throw($_message);} 

sub _nextQA 	{Kea::Object->throw($_message);} 
	
sub _nextDS 	{Kea::Object->throw($_message);}

sub _endOfFile 	{Kea::Object->throw($_message);}

################################################################################
	
#///////////////////////////////////////////////////////////////////////////////
#///////////////////////////////////////////////////////////////////////////////
1;

=pod
New Ace File Format from David Gordon at gordon@genome.washington.edu,
who is the author of Consed.

Refer to the accompanying sample_ace_file.txt (below)

AS <number of contigs> <total number of reads in ace file>

CO <contig name> <# of bases> <# of reads in contig> <# of base segments in
contig> <U or C>

The U or C indicates whether the contig has been complemented from the
way phrap originally created it.  Thus this is always U for an ace
file created by phrap.

BQ
This starts the list of base qualities for the unpadded consensus
bases.  The contig is the one from the previous CO, hence no name is
needed here.

AF <read name> <C or U> <padded start consensus position>

This line replaces the "AssembledFrom*" line in the previous ace file format.
C or U means complemented or uncomplemented.  The <read name> is the true
read name (no .comp on it as with the previous ace file format.) 

BS <padded start consensus position> <padded end consensus position> <read name>

This replaces the "BaseSegment*" line from the previous ace file format.

RD <read name> <# of padded bases> <# of whole read info items> <# of read tags>

QA <qual clipping start> <qual clipping end> <align clipping start> <align
clipping end>

This is new information not found in the previous ace file.  If the entire read
is low quality, then <qual clipping start> and <qual clipping end> will
both be -1.

DS CHROMAT_FILE: <name of chromat file> PHD_FILE: <name of phd file> TIME:
<date/time of the phd file>

This is replaces the DESCRIPTION line from the old ace file.

The following is for whole read info items.  These are not fully
implemented, and the format may eventually change.  The read is
implied by the location of the whole read info item within the ace
file.  They are found after the DS line for a read.

WR {
<tag type> <what program created tag> <date when tag was created in form YYMMDD>
}

The following is for transient read tags (those generated by
crossmatch and phrap).  They are not fully implemented, and the format
may eventually change.  The read is
implied by the location of the whole read info item within the ace
file.  They are found after the WR lines for a read.

RT{
<tag type> <what program created tag> <padded cons pos start> <padded cons pos
end> <date when tag was created in form YYMMDD>
}

There are consensus tags now in the ace file.  All consensus tags have
the following format:

CT{
<contig name> <tag type> <what program created tag> <padded cons pos start>
<padded cons pos end> <date when tag was created in form YYMMDD>
(possibly additional information)
}


In the case of most consensus tag types, there is only 1 line for the
consensus tag.  In the case of comment tags and oligo tags, there are
additional lines of information.  The comment tag includes the comment
on the additional lines.  The oligo tag has the following information:
<oligo name> <oligo bases from 5' to 3'> <melting temp> <C or U
indicating whether the oligo is top strand or bottom strand relative
to the orientation of the contig as created by phrap>

=cut